src/tests/ftest/launch.py

```diff
@@ -20,164 +20,471 @@
+
+    Args:
+        test_list (list): list of test scripts to run
+        machine_mapping (dict): the mapping of machine replacement keys to
+            their replacement values
+        debug (bool): whether or not to dispaly debug messages
+
+    Returns:
+        list: a list of dictionaries of each test script and yaml file
+
+    """
+    test_files = [{"py": test, "yaml": None} for test in test_list]
+    home_tmp = os.path.join(os.path.expanduser("~"), "tmp")
+    get_output("mkdir -p {}".format(home_tmp))
+    for test_file in test_files:
+        base, _ = os.path.splitext(test_file["py"])
+        file_name = os.path.basename(base)
+        if os.path.exists("{}.yaml".format(base)):
+            # Read the yaml file
+            display("Reading {}".format("{}.yaml".format(base)), debug)
+            with open("{}.yaml".format(base), "r") as yaml_file:
+                data = yaml_file.readlines()
+
+            # Replace the 'boro-[A-Z]' names with the specified machine names.
+            # Exclude any machine name placeholders that do not have a mapping.
+            tmp_yaml = os.path.join(home_tmp, "{}.yaml".format(file_name))
+            display("Creating {}".format(tmp_yaml), debug)
+            with open(tmp_yaml, "w") as yaml_file:
+                for line in data:
+                    match = re.findall(r"(.*- )(boro-[A-Z])", line)
+                    if len(match) > 0:
+                        replace_machine = match[0][1]
+                        display(
+                            "    MATCH: {} -> {}".format(
+                                line.rstrip(), replace_machine),
+                            debug)
+                        if replace_machine in machine_mapping.keys():
+                            new_line = "{}{}\n".format(
+                                match[0][0], machine_mapping[replace_machine])
+                            display(
+                                "  Line: {}".format(new_line.rstrip()),
+                                debug)
+                            yaml_file.write(new_line)
+                    else:
+                        display("  Line: {}".format(line.rstrip()), debug)
+                        yaml_file.write(line)
+            test_file["yaml"] = tmp_yaml
+
         else:
-            remaining_args = 2
-        addnl_tests = sys.argv[remaining_args:]
-    else:
-        printhelp()
+            display(
+                "ERROR: No yaml file exists for {}; excluding".format(
+                    test_file["py"]))
+            test_file["yaml"] = None
+    return test_files
 
-    # make it easy to specify a directory as a parameter later
-    test_directory = os.getcwd()
 
-    # setup some aspects of the environment
-    with open(os.path.join(os.path.dirname(os.path.realpath(__file__)),
-                           "../../../.build_vars.json")) as f:
-        envdata = json.load(f)
+def map_machines(machines):
```

Comment:
When run manually (outside of CI) I find it very useful to be able to apply the ftest.sh-like approach of updating the nodes in the yaml file with the hosts I have reserved.  I say ftesh.sh-like approach as this function accounts for yaml files that specify more host placeholders than may be reserved by the user.  Some tests can be run with less than the number of boro-* markers in the yaml file.  Currently, if a yaml file specifies more boro-* entries then hosts passed on the launch.py command line, those yaml entries are excluded in the temporary yaml file and we attempt to run the test with less servers/clients.  There will undoubtedly be situations were tests may not make sense to run with less hosts, which should result in a failed test.  Tests themselves could be more robust to handle this, or there could be an additional argument to launch.py that would not run a test if there were not enough server/clients to satisfy the yaml file.  I would not recommend running a test with non-replaced boro-* yaml entries.  Thoughts?

You also touched upon another interesting topic.  Does it make sense for ftest.sh or launch.py to handle the yaml boro-* replacements?  Launch.py only needs to change the yaml files that are needed for the specified tags.

Reply:
I do also want to split the `--machines` command line argument into `--servers` and `--clients` which would no longer mean that the second host specified would be a 1-to-1 replacement for `boro-B`.  The replacement code would end up taking into effect if the yaml entry was a `test_servers` or `test_clients` or `test_machines` entry.

Reply:
Yes, all really good questions/points.  I honestly don't care if the machine substituting is in `ftest.sh` or `launch.py`.  Certainly with the exception of the lack of cleaning up of the temporary `.yaml` files, I think your implementation is or at least will be more robust.

But ultimately, if/when we are going to land it, it ought to be used when it lands, as least as a proof of the new code.  Maybe it's not really ready to be used and you want to flesh it out some more first.  That's fine, but if it won't be used in this patch, it should probably be deferred to another patch and removed from this one.  In fact I would probably support that option in any case since it does seem to be a case of feaping creaturism for a patch that is intended to just support adding multiple tags.